src/server/injected/componentUtils.ts

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type ParsedComponentAttribute = {
+  jsonPath: string[],
+  op: '='|'*=',
+  value: any,
+  caseSensetive: boolean,
+};
+type ParsedComponentSelector = {
+  name: string,
+  attributes: ParsedComponentAttribute[],
+};
+
+export function checkComponentAttribute(obj: any, attr: ParsedComponentAttribute) {
+  for (const token of attr.jsonPath) {
+    if (obj)
+      obj = obj[token];
+  }
+  if (typeof obj !== typeof attr.value)
+    return false;
+  if (typeof obj === 'string') {
+    if (attr.op === '=')
+      return attr.caseSensetive ? obj === attr.value : obj.toUpperCase() === attr.value.toUpperCase();
+    if (attr.op === '*=')
+      return attr.caseSensetive ? obj.includes(attr.value) : obj.toUpperCase().includes(attr.value);
+  }
+  if (attr.op === '=')
+    return obj === attr.value;
+  return false;
+}
+
+export function parseComponentSelector(selector: string): ParsedComponentSelector {
+  let wp = 0;
+  let EOL = selector.length === 0;
+  let next = selector.length ? selector[0] : '';
+  const eat1 = () => {
+    const result = next;
+    ++wp;
+    next = wp < selector.length ? selector[wp] : '';
+    EOL = wp >= selector.length;
+    return result;
+  };
+  const syntaxError = (name: string) => {
+    if (EOL)
+      throw new Error(`Unexpected end of selector \`${selector}\` while parsing "${name}"`);
+    throw new Error(`Error while parsing selector \`${selector}\` - unexpected symbol "${next}" at position ${wp} while parsing ${name}`);
+  };
+
+  function skipSpaces() {
+    while (!EOL && /\s/.test(next)) eat1();
+  }
+
+  function readIdentifier() {
+    let result = '';
+    while (!EOL && /[-$0-9A-Z_]/i.test(next))
+      result += eat1();
+    return result;
+  }
+
+  function readQuotedString(quote: string) {
+    let result = eat1();
+    if (result !== quote)
+      syntaxError('quoted string');
+    while (!EOL && next !== quote)
+      result += eat1();
+    if (next !== quote)
+      syntaxError('quoted string');
+    result += eat1();
+    return result;
+  }
+
+  function readAttributeToken() {
+    skipSpaces();
+    let token = '';
+    if (next === `'` || next === `"`)
+      token = readQuotedString(next).slice(1, -1);
+    else
+      token = readIdentifier();
+    skipSpaces();
```

Comment:
It makes for an easier parser if you always skip spaces at the start of a token, and never at the end. Or you can move all skipSpaces calls to readAttribute() below.

Reply:
Aligned skipSpaces next to every token (e.g. `next` reading)